### PR TITLE
cargo-llvm-cov: init at 0.4.13

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Closes #170938
Supersedes #170938

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by DieracDelta — APPROVED** [review]
LGTM, the fetchcrate seems like it fixes ofborg. Excited to see this in nixpkgs!

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (FliegendeWurst) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:19]: To fix running the tests and to fix running `cargo llvm-cov` without a rustup toolchain installation:
```suggestion
{ lib, rustPlatform, fetchCrate, fetchFromGitHub, libllvm, makeBinaryWrapper }:

rustPlatform.buildRustPackage rec {
  pname = "cargo-llvm-cov";
  version = "0.4.13";

  # required because upstream does not keep a Cargo.lock file in their repository
  # (and some of the files required for running the tests are not published to crates.io)
  src = fetchFromGitHub {
    owner = "taiki-e";
    repo = "cargo-llvm-cov";
    rev = "v${version}";
    sha256 = "sha256-DfkBBPlJhxKeH6LZPAm3AQfqwNWTTSlxANUxMTGha28=";
  };
  srcCrate = fetchCrate {
    inherit pname version;
    sha256 = "sha256-x1M1rvQXiNX1HUgBtCLJy2Pdkes8I0ucGpNSweT5Tfs=";
  };
  prePatch = ''
    cp ${srcCrate}/Cargo.lock .
  '';

  nativeBuildInputs = [ makeBinaryWrapper ];
  buildInputs = [ libllvm ];

  LLVM_COV = "${rustPlatform.rust.rustc.llvmPackages.llvm}/bin/llvm-cov";
  LLVM_PROFDATA = "${rustPlatform.rust.rustc.llvmPackages.llvm}/bin/llvm-profdata"; 

  postInstall = ''
    wrapProgram $out/bin/cargo-llvm-cov \
      --set LLVM_COV "${rustPlatform.rust.rustc.llvmPackages.llvm}/bin/llvm-cov" \
      --set LLVM_PROFDATA "${rustPlatform.rust.rustc.llvmPackages.llvm}/bin/llvm-profdata"
  '';

  cargoSha256 = "sha256-IOngZiyONbxPzpNdKE6UM9hlZcV4ZfKZdVPX9+vgqjI=";

  meta = with lib; {
    description = "Cargo subcommand to gather source-based code coverage using LLVM";
```
(may not be the most elegant solution)

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:19]:
```suggestion

  postPatch = ''
```

**Comment by matthiasbeyer** [comment]
@FliegendeWurst unfortunately, your workaround does not work anymore with 0.4.14... care to investigate?

**Review by winterqt — CHANGES_REQUESTED** [review]
What I'm confused about is that this should have never worked in the first place, for a few reasons:

1. Your hash for the crate source is wrong (off by one character).
2. Your patch for copying the `Cargo.lock` doesn't apply to the dependency tarball generation, which means the build still fails.
2. The tests depend on a nightly compiler, which we don't ship (and `rustPlatform` doesn't use).

Due to these reasons, I propose the following solution, which I have locally and can push if you'd like (or you can make the changes yourself):

1. Only use the crate source, we don't care about the tests fixtures
2. `rm test/tests.rs` in `postPatch`, to stop the tests that require nightly from running

Making these changes locally has this building successfully for me.

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:15]: The tests can still be ran, but only a subset of them:

```suggestion
  src = fetchCrate {
    inherit pname version;
    sha256 = "sha256-DY5eBSx/PSmKaG7I6scDEbyZQ5hknA/pfl0KjTNqZlo=";
  };

  postPatch = ''
    rm test/tests.rs
  '';

  nativeBuildInputs = [ makeBinaryWrapper ];
  buildInputs = [ libllvm ];
```

File: pkgs/top-level/all-packages.nix
Comment (winterqt) [inline pkgs/top-level/all-packages.nix:14197]: Nit: can you swap these so they're properly sorted?

**Review by winterqt — CHANGES_REQUESTED** [review]

**Comment by matthiasbeyer** [comment]
With your changes, this fails now with:

```
rm: cannot remove 'test/tests.rs': No such file or directory
```

**Comment by winterqt** [comment]
Ugh, sorry, made a typo in my original message and didn't notice before repeating it: it should be `rm tests/test.rs`.

**Comment by matthiasbeyer** [comment]
I'm gonna use fixup patches for now and squash once all of you approve.

**Review by FliegendeWurst — APPROVED** [review]

**Review by DieracDelta — APPROVED** [review]

**Comment by DieracDelta** [comment]
Any chance we get this merged? I would love to drop my overlays for this and pull directly from nixpkgs.

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:16]: `makeWrapper` is much more commonly used over `makeBinaryWrapper` for simple wrapping like this:
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:20]: What are these required for, if the resulting binary is wrapped? Does the build require this?

**Review by winterqt — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (FliegendeWurst) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:20]: They would be required for the tests (which are disabled at the moment).

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:16]: Guess `makeBinaryWrapper` is preferred, per manual. Wonder why we don't use it more...

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:17]: Should this use the same libllvm as rustc (e.g. `rustPlatform.rust.rustc.llvmPackages.libllvm`)?

**Comment by winterqt** [comment]
I'm happy to merge this after https://github.com/NixOS/nixpkgs/pull/185183#discussion_r1002818140 is figured out. I don't see why we shouldn't use the same version (smaller closure size and all), but maybe it matters?

Apologies for more or less forgetting about this PR.

**Comment by CobaltCause** [comment]
Is there any reason not to update this PR to init at upstream's 0.5.0?

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:17]: Not sure if this is different because it's an older version, but I didn't need to add any `buildInputs` at all to get 0.5.0 to build/function.

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (CobaltCause) [inline pkgs/development/tools/rust/cargo-llvm-cov/default.nix:25]: I don't think wrapping this is a good idea, since it prevents users from choosing their own version of the Rust toolchain. For example, if I'm writing a flake that uses `fenix`, `cargo-llvm-cov` wouldn't be able to see the version of `llvm-tools-preview` from it, which might cause problems due to version disparities.

Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:25]: Good point.

I wonder if just leaving it unwrapped is good UX, though -- will users know to set those environment variables intuitively?

Comment (CobaltCause) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:25]: The command prints out a pretty helpful error message in situations where the environment isn't set up right:

```
$ cargo llvm-cov
error: failed to find llvm-tools-preview, please install llvm-tools-preview, or set LLVM_COV and LLVM_PROFDATA environment variables
```

File: pkgs/development/tools/rust/cargo-llvm-cov/default.nix
Comment (winterqt) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:17]: Yeah, looks like libllvm was never required, at least on 0.4.13.

@DieracDelta Do you remember why you added it?

Comment (DieracDelta) [reply pkgs/development/tools/rust/cargo-llvm-cov/default.nix:17]: I have no recollection of adding this. If it works without libllvm, then let's drop the dependency 👍

**Comment by winterqt** [comment]
I'm going to close this in favor of #197478, because it has all the tests working, and does not wrap the binary.

Thanks to everyone who got this implementation where it was, and if you can see anything in there that's worth a review, don't hesitate to comment there :)

**Pull request closed without merging**
